脆弱性解消

keiz1213 opened the pull request:
1.SQLインジェクション対策
2.XSS対策
3.自分以外の人がブログの編集、更新、削除を行えないようにした
4.有料機能を無料ユーザーが使えなくした
5.他人の秘密ブログにアクセスできないようにした
6.APIの返却値から個人情報を削除

keiz1213 committed:
脆弱性解消

1.SQLインジェクション対策
2.XSS対策
3.自分以外の人がブログの編集、更新、削除を行えないようにした
4.有料機能を無料ユーザーが使えなくした
5.他人の秘密ブログにアクセスできないようにした
6.APIの返却値から個人情報を削除

keiz1213 committed:
correct_user修正

keiz1213 committed:
before_actionの位置を修正

keiz1213 committed:
自然な形でActiveRecord::RecordNotFoundが出るようにした

1. raiseでエラーを起こしていた箇所が不適切だと思い、修正

keiz1213 committed:
メソッド名修正

yoshitsugu commented on app/controllers/blogs_controller.rb:
```diff
@@ -48,6 +50,17 @@ def set_blog
+
+  def secret_blog_filter
+    blog = Blog.find(params[:id])
+    params[:id] = nil if blog.secret && !(user_signed_in? && blog.user == current_user)
```
params自体を変更してしまうと、もし今後paramsを他の用途に使いたい、というときに困ってしまいます。
Blogの検索条件を `where` などを使ってうまく絞り込んでみてください〜

keiz1213 replied:
なるほどです。こちら修正してみました。

yoshitsugu commented on app/controllers/blogs_controller.rb:
```diff
@@ -48,6 +50,17 @@ def set_blog
+    params.require(:blog).permit(*attributes)
+  end
+
+  def set_correct_blog
```
確かに `correct` なblogをとってきたいものではありますが `correct` かどうかは文脈でだいぶかわってしまうので、ここでは `my_blog` とか `current_users_blog` とかでどうでしょうか

keiz1213 replied:
確かに、これだと文脈で捉え方が曖昧になりそうです。修正しました。

yoshitsugu commented on app/controllers/blogs_controller.rb:
```diff
@@ -48,6 +50,17 @@ def set_blog
 
   def blog_params
-    params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+    attributes = %i[title content secret random_eyecatch]
+    attributes.delete(:random_eyecatch) unless current_user.premium
```
逆に初期値は `:random_eyecatch` はナシで、特定の条件のときだけつけるようにしたほうが事故が少ないかなーと思います！

keiz1213 replied:
なるほどです。確かに、random_eyecatchは初期値としてはじめから無いほうが安心ですね。こちら修正しました。勉強になります。

yoshitsugu commented on app/models/blog.rb:
```diff
@@ -10,7 +10,7 @@ class Blog < ApplicationRecord
 
   scope :search, lambda { |term|
-    where("title LIKE '%#{term}%' OR content LIKE '%#{term}%'")
+    where('title LIKE ? OR content LIKE ?', "%#{term}%", "%#{term}%")
```
今回はスコープ外なので、大丈夫なんですが、 `LIKE` 句を使うときは `sanitize_sql_like` というサニタイズを使うようにするとよいので、今後の参考までに。
https://api.rubyonrails.org/classes/ActiveRecord/Sanitization/ClassMethods.html#method-i-sanitize_sql_like

keiz1213 replied:
なるほどです。LIKE句用にそのようなサニタイズがあるのですね。調べ方が甘かったです😅覚えておきます✍️

yoshitsugu commented on app/views/api/liking_users/_user.json.jbuilder:
```diff
@@ -1,2 +1,2 @@
-json.extract! user, :id, :email, :created_at, :updated_at, :nickname, :premium
+json.extract! user, :id, :created_at, :updated_at, :nickname, :premium
```
そもそも、必要最低限のデータだけ返すようにするとより安全かなと思います。考えてみてください〜。

keiz1213 replied:
確かにです。こちら修正しました。

yoshitsugu reviewed (commented):
気になったところコメントしました

keiz1213 committed:
指摘事項修正

1. 秘密のブログへのアクセス制御の方法を変更
2. set_correct_blogを具体的に理解しやすくするため名前を変更
3. 安全性を高めるため、blog_paramsのrandom_eyecatchを初期値から削除し、条件に合う場合のみ、追加するようにした
4. APIの返却値を必要最低限の情報のみ返すようにした

yoshitsugu reviewed (approved):
OKです！
